### PR TITLE
refactor: use k8s_service_info lib instead of SDI

### LINKED ISSUE #413: update the grpc relation to use the mlops-libs `k8s_service_info` library
### Context

Previous versions of the `metadata-writer-grpc` use an SDI-backed implementation of the grpc relation with mlmd.  For the Charmed Kubeflow 1.9 release, mlmd's relation handling is changing to use the [mlops-libs k8s_service_info](https://github.com/canonical/mlops-libs/blob/main/lib/charms/mlops_libs/v0/k8s_service_info.py) library.  We need to update that here as well to keep them compatible.  

### What needs to get done

Update the relation handling here to use the [mlops-libs k8s_service_info](https://github.com/canonical/mlops-libs/blob/main/lib/charms/mlops_libs/v0/k8s_service_info.py) library

### Definition of Done

charm has the new relation handling and is demonstrated working.  If this is implemented before mlmd is upgraded, we might need to demonstrate this using a dummy charm of an intermediate implementation of mlmd because we have no other charms that use this library yet.

### PR DESCRIPTION
**Pull request by DnPlas**
Use the k8s_service_info for receiving the MLMD GRPC Service info instead of using the SDI, as it will stop being supported soon.
This commit also ensures that mlmd runs with trust=True in the integration tests.

Fixes #413

NOTE for reviewers:

* I have added a component for wrapping the k8s service info requirer because we need to:
1. Have a way to return a status based on the relation conditions
2. Because there is no other way to do logic that prevents errors in the init method of the charm. For instance, it would be weird to have a try/catch in the init method for ensuring the required relation is established.

* Also note that this component could be placed in chisme, BUT that requires us to put charm libraries in the package, which is under discussion by the team.

* Depends on https://github.com/canonical/envoy-operator/pull/87

#### Testing instructions

1. Deploy the charm in this branch
2. Deploy `mlmd` from `channel latest/edge` with `trust`
3. Relate both charms
4. Verify both charms go to active and idle

##### Test upgrade path
1. Deploy `kfp-metadata-writer` from `channel 2.0/stable` with `trust`
2. Run `juju refresh` to the charm in this branch
3. Observe the charm going to Blocked(because it is missing a relation with mlmd) with the messsage: `[grpc] Missing relation with a k8s service info provider. Please add the missing relation`

### CODE REVIEW

File: .github/workflows/integrate.yaml
Comment (ca-scribner) [inline .github/workflows/integrate.yaml:147]: Should this be separated into its own PR?  If this causes an issue, it'll be unexpected to find it here

File: charms/kfp-metadata-writer/requirements.in
Comment (ca-scribner) [inline charms/kfp-metadata-writer/requirements.in:9]: Can we specify the ubuntu version this applies to (20.04?)?  Is there something that'll get fixed that lets us remove this restriction, or an upstream bug to track?

File: charms/kfp-metadata-writer/src/charm.py
Comment (ca-scribner) [inline charms/kfp-metadata-writer/src/charm.py:51]: Not a hard requirement, but I'd include the relation events in the `K8sServiceInfoComponent` class directly like [the sdi component does](https://github.com/canonical/charmed-kubeflow-chisme/blob/8a7454fede8dbcefe65cd45f2fc04a3926b7e421/src/charmed_kubeflow_chisme/components/serialised_data_interface_components.py#L84).  Main reason is I don't think there'd be a use case for us to need this component and not observe the relation events

Comment (ca-scribner) [inline charms/kfp-metadata-writer/src/charm.py:51]: Do we also need `relation_broken` here?  I see it in the [SDI component](https://github.com/canonical/charmed-kubeflow-chisme/blob/8a7454fede8dbcefe65cd45f2fc04a3926b7e421/src/charmed_kubeflow_chisme/components/serialised_data_interface_components.py#L86).  I think yes, because if we remove the relation here the user wont be notified there is a problem

File: charms/kfp-metadata-writer/src/components/k8s_service_info_component.py
Comment (ca-scribner) [inline charms/kfp-metadata-writer/src/components/k8s_service_info_component.py:53]: I worry this is too harsh.  This will happen transiently when a relation is established [^1].  Can we add some explanation to this saying it can be transient, but if it persists then it is an error?

[^1]: because [`relation-changed` always follows a `relation-joined`, so there will be some time between relation-joined and when the other side actually posts their data (triggering a second `relation-changed`)

File: charms/kfp-metadata-writer/tests/unit/test_operator.py
Comment (ca-scribner) [inline charms/kfp-metadata-writer/tests/unit/test_operator.py:68]: I think we should remove this check.  This tests Juju/harness behaviour, not something we created.  the above test of `get_service_info() == MOCK_GRPC_DATA` completely tests your code

File: charms/kfp-metadata-writer/tests/unit/test_operator.py
Comment (ca-scribner) [inline charms/kfp-metadata-writer/tests/unit/test_operator.py:88]: Missing feels a bit generic (and could mean "empty" which is tested above).  Maybe `partial` ?
```suggestion
def test_grpc_relation_with_partial_data(harness, mocked_lightkube_client):
    """Test the grpc relation raises when data incomplete."""
```

**Review by ca-scribner — CHANGES_REQUESTED** [review]
A few changes suggested above

Is the CI failure expected?

Also, iiuc this will fail CI until after canonical/envoy-operator#87 right?

**Comment by ca-scribner** [comment]
Created canonical/charmed-kubeflow-chisme#94 to capture the discussion about how we could reuse code like the Component here that needs a charm library

File: charms/kfp-metadata-writer/requirements.in
Comment (ca-scribner) [inline charms/kfp-metadata-writer/requirements.in:9]: Not a bug exactly, but I noticed during testing that build times have increased by ~8min:
* [before](https://github.com/canonical/envoy-operator/actions/runs/8569177476):
  * publish: ~3-4min
  * integration: 14min
* [after](https://github.com/canonical/envoy-operator/actions/runs/8724143722): 
  * publish: ~11min
  * integration: 21min
  
I'm 99% certain this is because of the time taken to build pydantic.  I'm documenting this in the mlops repo

File: tests/integration/bundles/kfp_latest_edge.yaml.j2
Comment (ca-scribner) [inline tests/integration/bundles/kfp_latest_edge.yaml.j2:1]: I think the relation:
```yaml
- [envoy:grpc, mlmd:grpc]
```

needs to be updated to:
```yaml
- [envoy:grpc, mlmd:k8s-service-info]
```

because the mlmd sidecar rewrite [changed the relation name](https://github.com/canonical/mlmd-operator/pull/72/files#diff-d186204ea7abb85bbadd9d94ab29d5bcefcc43d6a0ed06015e98a71b2f099423L15)

Was the relation renaming in mlmd on purpose?  I wonder if we should revert that change instead and set the relation name back to `grpc` in mlmd

File: charms/kfp-metadata-writer/src/charm.py
Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py:51]: Yeah, we do.  Tested manually and if I break a relation, I only see these events in the logs:
```
unit-kfp-metadata-writer-0: 15:31:47 INFO juju.worker.uniter.operation ran "grpc-relation-departed" hook (via hook dispatching script: dispatch)
unit-kfp-metadata-writer-0: 15:31:48 INFO juju.worker.uniter.operation ran "grpc-relation-broken" hook (via hook dispatching script: dispatch)
```
Currently, the charm wont know that it has lost its grpc service

Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py:51]: Oh sorry no, this is a bug.

The `KubernetesServiceInfoRequirer` already [observes the relation events](https://github1s.com/canonical/kfp-operators/blob/KF-5485-use-k8s-service-info-lib/charms/kfp-metadata-writer/lib/charms/mlops_libs/v0/k8s_service_info.py#L192-L200) - we don't need to observe those events here.  But, the `K8sServiceInfoComponent` **does need to observe the library's [`KubernetesServiceInfoUpdatedEvent`](https://github1s.com/canonical/kfp-operators/blob/KF-5485-use-k8s-service-info-lib/charms/kfp-metadata-writer/lib/charms/mlops_libs/v0/k8s_service_info.py#L143-L144) so the charm knows to wake whenever the lib sees updated data.  

Right now on a relation-data-changed event the underlying lib is handling the event and emitting a `KubernetesServiceInfoUpdatedEvent`, but nothing in the charm observes that so the charm itself is not doing anything with the new relation data.  You can see this by watching the `juju debug-log -i kfp-metadata-writer` while relating to mlmd - nothing happens and the charm stays blocked.  The charm does eventually recover from this Blocked state though when the next `update_status` event fires (triggering a full charm reconcile).

Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py:51]: This bug can be caught by a unit test, so we must have a test coverage gap.

I think the issue is that:
* [this test](https://github.com/canonical/kfp-operators/pull/436/files#diff-8e7738d6ff50a7d31913bee76270ebc0c80efefbcf3c75164fb3521f7bb4603eR62) asserts the charm *can access* the relation data, not that it actually used it to reconcile the charm after the relation was established
* [this test](https://github.com/canonical/kfp-operators/pull/436/files#diff-8e7738d6ff50a7d31913bee76270ebc0c80efefbcf3c75164fb3521f7bb4603eR132) does show we successfully use the data, but it does not prove the relation-changed event reconciled the charm either because it emits an `install` event right after the data is updated (the install event then **uses** the data, masking how the relation-changed event did not use the data)

If we move the `harness.charm.on.install.emit()` in the second test above the `harness.add_relation`, we now see the test fails

Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py:51]: Test it yourself to be sure, but I think [this pr against your pr](https://github.com/canonical/kfp-operators/pull/437) will fix this issue

File: charms/kfp-metadata-writer/requirements.in
Comment (DnPlas) [reply charms/kfp-metadata-writer/requirements.in:9]: I'll add more information about the issue and when this pin can be removed.

File: charms/kfp-metadata-writer/src/charm.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py:51]: > but nothing in the charm observes that so the charm itself is not doing anything with the new relation data. You can see this by watching the juju debug-log -i kfp-metadata-writer while relating to mlmd - nothing happens and the charm stays blocked. The charm does eventually recover from this Blocked state though when the next update_status event fires (triggering a full charm reconcile).

yes, I observed exactly that behaviour when deploying the charm. I thought it was a limitation of the charmReconciler, though, as the library is in fact observing that exact event, and I kind of had to "force" it.

Maybe it's because I am less charmReconciler savvy, but it was not very obvious to me how to make this component observe non-standard events. I see [this guide](https://github.com/canonical/charmed-kubeflow-chisme/blob/main/src/charmed_kubeflow_chisme/components/how-to/authoring-a-simple-component.md#handling-non-standard-events), but I guess it needs some work.

I will make the proper changes (including the unit testing), and perhaps file an issue in chisme to improve the docs if that's okay.

Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py:51]: The library already handles that, I think what we need to do is observe both relation changed and relation broken events as suggested in your [previous comment](https://github.com/canonical/kfp-operators/pull/436#discussion_r1569449222).

File: charms/kfp-metadata-writer/requirements.in
Comment (DnPlas) [reply charms/kfp-metadata-writer/requirements.in:9]: Yes, that's the reason. Thanks for filing the issue, we can discuss this subject there. I don't think there is an action item in this PR atm, but let me know otherwise.

File: tests/integration/bundles/kfp_latest_edge.yaml.j2
Comment (DnPlas) [reply tests/integration/bundles/kfp_latest_edge.yaml.j2:1]: I have submitted a PR for reverting the name, I think it is okay to revert it as we won't have to change the name in the multiple places we have it.
As for the upgrade path, since `mlmd` was rewritten from podspec to sidecar pattern, running `juju refresh` on the charm won't be possible for upgrading it; users have to delete the charm and deploy a new one, thus existing relations will anyway be broken and will be removed when `mlmd` is gone.

File: charms/kfp-metadata-writer/src/charm.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py:51]: In your PR you are suggesting to remove the `refresh_event` arg from the `Class`. I think we should keep it to not limit the library's functionality, or use `*args`. wdyt?

File: charms/kfp-metadata-writer/src/components/k8s_service_info_component.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/components/k8s_service_info_component.py:53]: Should we return a `WaitingStatus`?

While I agree this could be too harsh, I worry that we report back a status when we should report back an error. I have seen in other charms we do this, and most of the times it is not very obvious what is going on. I'm open to suggestions.

File: charms/kfp-metadata-writer/requirements.in
Comment (DnPlas) [reply charms/kfp-metadata-writer/requirements.in]: Updated, wdyt?

File: charms/kfp-metadata-writer/src/charm.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py]: Updated, wdyt?

File: charms/kfp-metadata-writer/tests/unit/test_operator.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/tests/unit/test_operator.py]: Updated.

File: charms/kfp-metadata-writer/requirements.in
Comment (ca-scribner) [reply charms/kfp-metadata-writer/requirements.in]: lgtm! ty

File: charms/kfp-metadata-writer/src/charm.py
Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py]: Yeah sorry it is not really clear how to use this.  I'd also accept the criticism of "To set the events I'm observing, why am I overwriting the private attribute `self._events_to_observe`?" :D  It shouldn't be this way, we just ran out of time before and never came back to make this better.  

Your PR as of now lgtm for this.  I'm guessing you came around on removing the `refresh_event`?  To me, since this code will always get used only for this specific setup, leaving that arg out is ok

Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py]: The updates lgtm in terms of function, but I don't see the value keeping the `refresh_event` arg.  Are there events other than `relation_changed` and `relation_broken` that should trigger `KubernetesServiceInfoRequirer`?  I don't think there are.

Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py]: yeah I think I wrote this comment before [the above comment](https://github.com/canonical/kfp-operators/pull/436#discussion_r1569224916), and misunderstood why the `relation-broken` wasn't working.  We can close this

File: charms/kfp-metadata-writer/src/components/k8s_service_info_component.py
Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/components/k8s_service_info_component.py:55]: I suggest `WaitingStatus` with a message making it clear that, if `WaitingStatus` persists, its probably an error.  So like `WaitingStatus("No data found in X relation.  This may be transient, but if it persists it is likely an error") or similar

File: tests/integration/bundles/kfp_latest_edge.yaml.j2
Comment (ca-scribner) [reply tests/integration/bundles/kfp_latest_edge.yaml.j2:1]: k if we're reverting the name then this lgtm!

File: charms/kfp-metadata-writer/src/charm.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py]: Re `refresh_event` -  my intention is to keep the component as generic as possible, I see this being abstracted to chisme eventually, so I had it there. You do have a point and perhaps we don't really need it, at least for this charm we don't. I don't have a hard opinion on this one, though.

File: charms/kfp-metadata-writer/src/components/k8s_service_info_component.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/components/k8s_service_info_component.py:55]: Personally I do not like that message, I have seen many charms that use the charmReconciler hidding actual issues behind this message and for someone who has to debug charms, it is a pain to spot the actual issue. I can change it to what you are suggesting, but knowing this is a behaviour we should probably change.

File: charms/kfp-metadata-writer/src/charm.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py]: Updated, let me know what you think

File: charms/kfp-metadata-writer/src/components/k8s_service_info_component.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/components/k8s_service_info_component.py]: Updated, let me know what you think

File: charms/kfp-metadata-writer/src/charm.py
Comment (DnPlas) [reply charms/kfp-metadata-writer/src/charm.py]: @ca-scribner I have tested this in a live model and unfortunately, `relation-broken` is not observed(?). I will need some guidance on this. While adding the `self._events_to_observe` to observe the `relation-update` event, solves the issue with seeing the unit status go from blocked to active, removing the relation does not seem to trigger anything to go back from active to blocked.

**Comment by DnPlas** [comment]
After we have merged https://github.com/canonical/envoy-operator/pull/87, I expect the bundle integration tests (v1, and v2) to be blocked by https://github.com/canonical/charmcraft/issues/1659. To mitigate this error, we could build `kfp-metadata-writer` at the end of the list to ensure no other charm is running `apt` processes. I have tested this workaround in #438.

File: charms/kfp-metadata-writer/src/charm.py
Comment (ca-scribner) [reply charms/kfp-metadata-writer/src/charm.py]: I think this makes more sense.  But if you do reuse it and have to add `refresh_event` back in, I grant you the right to say "I told you so" :)

**Review by ca-scribner — DISMISSED** [review]
the change itself here lgtm.  Sorting out the CI failures likely needs some more tweaks as @DnPlas mentions though.

Looked like there's also some failures in kfp-api and other integration test CI, but they might be flakes?

**Comment by DnPlas** [comment]
For the CI failures, please see [this comment](https://github.com/canonical/kfp-operators/pull/441#issuecomment-2069022059).

**Comment by DnPlas** [comment]
The CI should now be green.

**Review by ca-scribner — APPROVED** [review]

**Pull request merged**
